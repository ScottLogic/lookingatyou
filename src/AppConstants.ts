export enum EyeSide {
    RIGHT = 'RIGHT',
    LEFT = 'LEFT',
}
export const eyelidPosition = {
    OPEN: 0.45,
    CLOSED: 0,
    SQUINT: 0.2,
    SHOCKED: 0.75,
};
export const pupilSizes = {
    dilated: 1.3,
    neutral: 1.0,
    constricted: 0.8,
};
export const pupilSizeChangeInterval = 2500;
export const middleX = 0;
export const middleY = 0;
export const xIncrement = 1;
export const moveSize = 0.4;
export const buffer = 0.2;
export const maxBrightness = 220;
export const dilationMultipler = 0.8;
export const dilationOffset = 0.7;
export const sleepDelay = 25000;
export const transitionTime = {
    blink: 200,
    dilate: 750,
};
export const neutralBlinkFrequency = 0.25;
export const irisSkewFactor = 0.8;
export const maxMoveWithoutBlink = 0.8;
export const chosenTargetColour = 'green';
export const nonChosenTargetColour = 'red';
export const maxNumTargetsToConsider = 5;
export const bodyParts = {
    RIGHT_SHOULDER: 'rightShoulder',
    LEFT_SHOULDER: 'leftShoulder',
    RIGHT_HIP: 'rightHip',
};
