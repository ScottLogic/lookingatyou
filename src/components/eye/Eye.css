.Eyelids path {
    fill: black;
    stroke: black;
    stroke-width: 2;
}

.BlackFill path {
    fill: black;
    stroke: black;
    stroke-width: 2;
}

.inner,
.outerReflection,
.innerReflection {
    transform-origin: center;
}
