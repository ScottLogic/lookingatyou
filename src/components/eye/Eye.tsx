import React from 'react';
import { eyes, transitionTime } from '../../AppConstants';
import './Eye.css';

interface IEyeProps {
    class: string;
    width: number;
    height: number;
    irisColor: string;
    scleraRadius: number;
    irisRadius: number;
    pupilRadius: number;
    openCoefficient: number;
    dilatedCoefficient: number;
    innerX: number;
    innerY: number;
    fps: number;
}

export default class Eye extends React.Component<IEyeProps> {
    private circleTransitionStyle: { transition: string };
    private eyelidTransitionStyle: { transition: string };
    constructor(props: IEyeProps) {
        super(props);
        this.circleTransitionStyle = {
            transition: `r ${transitionTime.dilate}ms, cx ${1000 /
                props.fps}ms, cy ${1000 / props.fps}ms`, // cx and cy transitions based on FPS
        };
        this.eyelidTransitionStyle = {
            transition: `d ${transitionTime.blink}ms`,
        };
    }

    renderCircle(
        radius: number,
        name: string,
        colour: string,
        centerX: number = this.props.width / 2,
        centerY: number = this.props.height / 2,
    ) {
        return (
            <circle
                style={this.circleTransitionStyle}
                r={radius}
                className={name}
                fill={colour}
                cx={centerX}
                cy={centerY}
            />
        );
    }

    render() {
        const eyeMiddleX = this.props.width / 2;
        const eyeLeft = eyeMiddleX - this.props.scleraRadius;
        const eyeRight = eyeMiddleX + this.props.scleraRadius;
        const eyeMiddleY = this.props.height / 2;

        const topEyelidY =
            eyeMiddleY - this.props.scleraRadius * this.props.openCoefficient;
        const bottomEyelidY =
            eyeMiddleY + this.props.scleraRadius * this.props.openCoefficient;

        const bezierCurveConstant = 0.55228474983; // (4/3)tan(pi/8)
        const bezierControlOffset =
            this.props.scleraRadius * bezierCurveConstant;
        const scaledYBezierControlOffset =
            bezierControlOffset * this.props.openCoefficient;
        const scaledXBezierControlOffset =
            bezierControlOffset - scaledYBezierControlOffset;
        const innerTopCoefficient = 1.25;
        const innerBottomCoefficient = 1.1;
        const outerTopCoefficient = 0.85;
        const outerBottomCoefficient = 0.9;
        let leftTopCoefficient;
        let rightTopCoefficient;
        let leftBottomCoefficient;
        let rightBottomCoefficient;
        if (this.props.class === eyes.RIGHT) {
            leftTopCoefficient = innerTopCoefficient;
            rightTopCoefficient = outerTopCoefficient;
            leftBottomCoefficient = innerBottomCoefficient;
            rightBottomCoefficient = outerBottomCoefficient;
        } else {
            leftTopCoefficient = outerTopCoefficient;
            rightTopCoefficient = innerTopCoefficient;
            leftBottomCoefficient = outerBottomCoefficient;
            rightBottomCoefficient = innerBottomCoefficient;
        }
        return (
            <svg
                className={this.props.class}
                width={this.props.width}
                height={this.props.height}
            >
                {this.renderCircle(this.props.scleraRadius, 'sclera', 'white')}
                <g className="inner">
                    {this.renderCircle(
                        this.props.irisRadius,
                        'iris',
                        this.props.irisColor,
                        this.props.innerX,
                        this.props.innerY,
                    )}
                    {this.renderCircle(
                        this.props.pupilRadius * this.props.dilatedCoefficient,
                        'pupil',
                        'black',
                        this.props.innerX,
                        this.props.innerY,
                    )}
                </g>
                <svg className="Eyelids">
                    <path
                        style={this.eyelidTransitionStyle}
                        d={
                            // upper eyelid
                            `M ${eyeLeft} ${eyeMiddleY},
                         A ${this.props.scleraRadius} ${
                                this.props.scleraRadius
                            } 0 0 1 ${eyeRight} ${eyeMiddleY}
                         C ${eyeRight -
                             rightTopCoefficient *
                                 scaledXBezierControlOffset} ${eyeMiddleY -
                                scaledYBezierControlOffset}, 
                         ${eyeMiddleX +
                             bezierControlOffset} ${topEyelidY}, ${eyeMiddleX} ${topEyelidY}
                         C ${eyeMiddleX -
                             bezierControlOffset} ${topEyelidY}, ${eyeLeft +
                                leftTopCoefficient *
                                    scaledXBezierControlOffset} ${eyeMiddleY -
                                scaledYBezierControlOffset}, ${eyeLeft} ${eyeMiddleY}`
                        }
                    />
                    <path
                        style={this.eyelidTransitionStyle}
                        d={
                            // lower eyelid
                            `M ${eyeLeft} ${eyeMiddleY},
                         A ${this.props.scleraRadius} ${
                                this.props.scleraRadius
                            } 0 0 0 ${eyeRight} ${eyeMiddleY}
                         C ${eyeRight -
                             rightBottomCoefficient *
                                 scaledXBezierControlOffset} ${eyeMiddleY +
                                scaledYBezierControlOffset}, 
                         ${eyeMiddleX +
                             bezierControlOffset} ${bottomEyelidY}, ${eyeMiddleX} ${bottomEyelidY}
                         C ${eyeMiddleX -
                             bezierControlOffset} ${bottomEyelidY}, ${eyeLeft +
                                leftBottomCoefficient *
                                    scaledXBezierControlOffset} ${eyeMiddleY +
                                scaledYBezierControlOffset}, ${eyeLeft} ${eyeMiddleY}`
                        }
                    />
                </svg>
                <svg className="BlackFill">
                    <path
                        d={
                            // upper eyelid
                            `M ${0} ${eyeMiddleY},
                         L ${eyeLeft} ${eyeMiddleY},
                         A ${this.props.scleraRadius} ${
                                this.props.scleraRadius
                            } 0 0 1 ${eyeRight} ${eyeMiddleY}
                         L ${this.props.width} ${eyeMiddleY},
                         L ${this.props.width} 0
                         L 0 0`
                        }
                    />
                    <path
                        d={
                            // upper eyelid
                            `M ${0} ${eyeMiddleY},
                         L ${eyeLeft} ${eyeMiddleY},
                         A ${this.props.scleraRadius} ${
                                this.props.scleraRadius
                            } 0 0 0 ${eyeRight} ${eyeMiddleY}
                         L ${this.props.width} ${eyeMiddleY},
                         L ${this.props.width} ${this.props.height}
                         L 0 ${this.props.height}`
                        }
                    />
                </svg>
            </svg>
        );
    }
}
