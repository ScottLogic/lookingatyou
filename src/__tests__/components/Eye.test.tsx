import { shallow } from 'enzyme';
import React from 'react';
import { EyeSide } from '../../AppConstants';
import Eye, { IEyeProps } from '../../components/eye/Eye';

let props: IEyeProps;

describe('Eye', () => {
    beforeEach(() => {
        props = {
            class: EyeSide.LEFT,
            width: 500,
            height: 500,
            irisColor: 'blue',
            scleraRadius: 100,
            irisRadius: 50,
            pupilRadius: 20,
            openCoefficient: 0.5,
            dilatedCoefficient: 1,
            innerX: 250,
            innerY: 250,
            fps: 10,
        };
    });

    it('should render correctly', () => {
        const wrapper = shallow(<Eye {...props} />).debug();
        expect(wrapper).toMatchSnapshot();
    });
});
