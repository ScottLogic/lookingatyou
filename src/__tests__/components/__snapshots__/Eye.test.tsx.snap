// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Eye should render correctly 1`] = `
"<svg className=\\"eye\\" width={500} height={500}>
  <circle className=\\"sclera\\" style={{...}} r={100} fill=\\"url(#scleraGradient)\\" cx={250} cy={250} />
  <g className=\\"inner\\" style={{...}} transform=\\"\\\\n                scale(1, 1)\\\\n                skewX(0)\\\\n                \\">
    <circle className=\\"iris\\" style={{...}} r={50} fill=\\"url(#irisGradient)\\" cx={250} cy={250} />
    <g className=\\"irisStyling\\">
      <path d=\\"M 250 250 l 80 0\\\\nl -70.00000000000001 0\\\\nl 69.84218750000001 5.0234375\\\\nl -69.86197916666667 -4.3953125\\\\nl 69.38906250000001 9.398958333333333\\\\nl -69.44791666666667 -8.7734375\\\\nl 68.66197916666667 13.736979166666668\\\\nl -68.75989583333335 -13.116666666666669\\\\nl 67.66354166666666 18.02135416666667\\\\nl -67.80104166666668 -17.408333333333335\\\\nl 66.3984375 22.234375\\\\nl -66.57395833333334 -21.63125\\\\nl 64.87135416666666 26.359895833333333\\\\nl -65.08437500000001 -25.76875\\\\nl 63.08854166666667 30.38125\\\\nl -63.33802083333334 -29.804687500000004\\\\nl 61.05625 34.28229166666667\\\\nl -61.34166666666667 -33.72291666666667\\\\nl 58.78333333333334 38.0484375\\\\nl -59.103125 -37.5078125\\\\nl 56.278125 41.66458333333333\\\\nl -56.63125 -41.14479166666667\\\\nl 53.55104166666667 45.11614583333334\\\\nl -53.9359375 -44.61979166666667\\\\nl 50.612500000000004 48.389583333333334\\\\nl -51.02760416666667 -47.91822916666667\\\\nl 47.47395833333334 51.471875\\\\nl -47.91822916666667 -51.02760416666667\\\\nl 44.1484375 54.35156250000001\\\\nl -44.61979166666667 -53.9359375\\\\nl 40.64843750000001 57.01614583333334\\\\nl -41.14479166666667 -56.63125\\\\nl 36.98854166666667 59.456250000000004\\\\nl -37.5078125 -59.103125\\\\nl 33.18177083333334 61.661458333333336\\\\nl -33.72291666666667 -61.34166666666667\\\\nl 29.244791666666668 63.622916666666676\\\\nl -29.804687500000004 -63.33802083333334\\\\nl 25.192187500000003 65.33385416666667\\\\nl -25.76875 -65.08437500000001\\\\nl 21.040104166666666 66.78697916666667\\\\nl -21.63125 -66.57395833333334\\\\nl 16.805208333333333 67.97604166666667\\\\nl -17.408333333333335 -67.80104166666668\\\\nl 12.503645833333335 68.89739583333333\\\\nl -13.116666666666669 -68.75989583333335\\\\nl 8.152604166666668 69.54635416666667\\\\nl -8.7734375 -69.44791666666667\\\\nl 3.7697916666666673 69.92083333333333\\\\nl -4.3953125 -69.86197916666667\\\\nl -0.628125 70.01979166666666\\\\nl 0 -70.00000000000001\\\\nl -5.0234375 69.84218750000001\\\\nl 4.3953125 -69.86197916666667\\\\nl -9.398958333333333 69.38906250000001\\\\nl 8.7734375 -69.44791666666667\\\\nl -13.736979166666668 68.66197916666667\\\\nl 13.116666666666669 -68.75989583333335\\\\nl -18.02135416666667 67.66354166666666\\\\nl 17.408333333333335 -67.80104166666668\\\\nl -22.234375 66.3984375\\\\nl 21.63125 -66.57395833333334\\\\nl -26.359895833333333 64.87135416666666\\\\nl 25.76875 -65.08437500000001\\\\nl -30.38125 63.08854166666667\\\\nl 29.804687500000004 -63.33802083333334\\\\nl -34.28229166666667 61.05625\\\\nl 33.72291666666667 -61.34166666666667\\\\nl -38.0484375 58.78333333333334\\\\nl 37.5078125 -59.103125\\\\nl -41.66458333333333 56.278125\\\\nl 41.14479166666667 -56.63125\\\\nl -45.11614583333334 53.55104166666667\\\\nl 44.61979166666667 -53.9359375\\\\nl -48.389583333333334 50.612500000000004\\\\nl 47.91822916666667 -51.02760416666667\\\\nl -51.471875 47.47395833333334\\\\nl 51.02760416666667 -47.91822916666667\\\\nl -54.35156250000001 44.1484375\\\\nl 53.9359375 -44.61979166666667\\\\nl -57.01614583333334 40.64843750000001\\\\nl 56.63125 -41.14479166666667\\\\nl -59.456250000000004 36.98854166666667\\\\nl 59.103125 -37.5078125\\\\nl -61.661458333333336 33.18177083333334\\\\nl 61.34166666666667 -33.72291666666667\\\\nl -63.622916666666676 29.244791666666668\\\\nl 63.33802083333334 -29.804687500000004\\\\nl -65.33385416666667 25.192187500000003\\\\nl 65.08437500000001 -25.76875\\\\nl -66.78697916666667 21.040104166666666\\\\nl 66.57395833333334 -21.63125\\\\nl -67.97604166666667 16.805208333333333\\\\nl 67.80104166666668 -17.408333333333335\\\\nl -68.89739583333333 12.503645833333335\\\\nl 68.75989583333335 -13.116666666666669\\\\nl -69.54635416666667 8.152604166666668\\\\nl 69.44791666666667 -8.7734375\\\\nl -69.92083333333333 3.7697916666666673\\\\nl 69.86197916666667 -4.3953125\\\\nl -70.01979166666666 -0.628125\\\\nl 70.00000000000001 0\\\\nl -69.84218750000001 -5.0234375\\\\nl 69.86197916666667 4.3953125\\\\nl -69.38906250000001 -9.398958333333333\\\\nl 69.44791666666667 8.7734375\\\\nl -68.66197916666667 -13.736979166666668\\\\nl 68.75989583333335 13.116666666666669\\\\nl -67.66354166666666 -18.02135416666667\\\\nl 67.80104166666668 17.408333333333335\\\\nl -66.3984375 -22.234375\\\\nl 66.57395833333334 21.63125\\\\nl -64.87135416666666 -26.359895833333333\\\\nl 65.08437500000001 25.76875\\\\nl -63.08854166666667 -30.38125\\\\nl 63.33802083333334 29.804687500000004\\\\nl -61.05625 -34.28229166666667\\\\nl 61.34166666666667 33.72291666666667\\\\nl -58.78333333333334 -38.0484375\\\\nl 59.103125 37.5078125\\\\nl -56.278125 -41.66458333333333\\\\nl 56.63125 41.14479166666667\\\\nl -53.55104166666667 -45.11614583333334\\\\nl 53.9359375 44.61979166666667\\\\nl -50.612500000000004 -48.389583333333334\\\\nl 51.02760416666667 47.91822916666667\\\\nl -47.47395833333334 -51.471875\\\\nl 47.91822916666667 51.02760416666667\\\\nl -44.1484375 -54.35156250000001\\\\nl 44.61979166666667 53.9359375\\\\nl -40.64843750000001 -57.01614583333334\\\\nl 41.14479166666667 56.63125\\\\nl -36.98854166666667 -59.456250000000004\\\\nl 37.5078125 59.103125\\\\nl -33.18177083333334 -61.661458333333336\\\\nl 33.72291666666667 61.34166666666667\\\\nl -29.244791666666668 -63.622916666666676\\\\nl 29.804687500000004 63.33802083333334\\\\nl -25.192187500000003 -65.33385416666667\\\\nl 25.76875 65.08437500000001\\\\nl -21.040104166666666 -66.78697916666667\\\\nl 21.63125 66.57395833333334\\\\nl -16.805208333333333 -67.97604166666667\\\\nl 17.408333333333335 67.80104166666668\\\\nl -12.503645833333335 -68.89739583333333\\\\nl 13.116666666666669 68.75989583333335\\\\nl -8.152604166666668 -69.54635416666667\\\\nl 8.7734375 69.44791666666667\\\\nl -3.7697916666666673 -69.92083333333333\\\\nl 4.3953125 69.86197916666667\\\\nl 0.628125 -70.01979166666666\\\\nl 0 70.00000000000001\\\\nl 5.0234375 -69.84218750000001\\\\nl -4.3953125 69.86197916666667\\\\nl 9.398958333333333 -69.38906250000001\\\\nl -8.7734375 69.44791666666667\\\\nl 13.736979166666668 -68.66197916666667\\\\nl -13.116666666666669 68.75989583333335\\\\nl 18.02135416666667 -67.66354166666666\\\\nl -17.408333333333335 67.80104166666668\\\\nl 22.234375 -66.3984375\\\\nl -21.63125 66.57395833333334\\\\nl 26.359895833333333 -64.87135416666666\\\\nl -25.76875 65.08437500000001\\\\nl 30.38125 -63.08854166666667\\\\nl -29.804687500000004 63.33802083333334\\\\nl 34.28229166666667 -61.05625\\\\nl -33.72291666666667 61.34166666666667\\\\nl 38.0484375 -58.78333333333334\\\\nl -37.5078125 59.103125\\\\nl 41.66458333333333 -56.278125\\\\nl -41.14479166666667 56.63125\\\\nl 45.11614583333334 -53.55104166666667\\\\nl -44.61979166666667 53.9359375\\\\nl 48.389583333333334 -50.612500000000004\\\\nl -47.91822916666667 51.02760416666667\\\\nl 51.471875 -47.47395833333334\\\\nl -51.02760416666667 47.91822916666667\\\\nl 54.35156250000001 -44.1484375\\\\nl -53.9359375 44.61979166666667\\\\nl 57.01614583333334 -40.64843750000001\\\\nl -56.63125 41.14479166666667\\\\nl 59.456250000000004 -36.98854166666667\\\\nl -59.103125 37.5078125\\\\nl 61.661458333333336 -33.18177083333334\\\\nl -61.34166666666667 33.72291666666667\\\\nl 63.622916666666676 -29.244791666666668\\\\nl -63.33802083333334 29.804687500000004\\\\nl 65.33385416666667 -25.192187500000003\\\\nl -65.08437500000001 25.76875\\\\nl 66.78697916666667 -21.040104166666666\\\\nl -66.57395833333334 21.63125\\\\nl 67.97604166666667 -16.805208333333333\\\\nl -67.80104166666668 17.408333333333335\\\\nl 68.89739583333333 -12.503645833333335\\\\nl -68.75989583333335 13.116666666666669\\\\nl 69.54635416666667 -8.152604166666668\\\\nl -69.44791666666667 8.7734375\\\\nl 69.92083333333333 -3.7697916666666673\\\\nl -69.86197916666667 4.3953125\\\\nl 70.01979166666666 0.628125\\\\nl -70.00000000000001 0\\\\n\\" fill=\\"#0000cc\\" style={{...}} />
    </g>
    <circle className=\\"pupil\\" style={{...}} r={20} fill=\\"black\\" cx={250} cy={250} />
    <ellipse className=\\"innerReflection\\" style={{...}} rx={15} ry={7.5} fill=\\"url(#reflectionGradient)\\" cx={250} cy={270} transform=\\"rotate(135)\\" />
    <ellipse className=\\"outerReflection\\" style={{...}} rx={20} ry={10} fill=\\"url(#reflectionGradient)\\" cx={250} cy={300} transform=\\"rotate(135)\\" />
  </g>
  <svg className=\\"Eyelids\\">
    <path style={{...}} filter=\\"url(#shadowTop)\\" d=\\"M 150 250,\\\\n                         A 100 100 0 0 1 350 250\\\\n                         C 309.959355637325 222.3857625085,\\\\n                         305.228474983 200, 250 200\\\\n                         C 194.771525017 200, 169.32996624405 222.3857625085, 150 250\\" />
    <path style={{...}} filter=\\"url(#shadowBottom)\\" d=\\"M 150 250,\\\\n                         A 100 100 0 0 0 350 250\\\\n                         C 319.62433875935 277.6142374915,\\\\n                         305.228474983 300, 250 300\\\\n                         C 194.771525017 300, 163.80711874575 277.6142374915, 150 250\\" />
  </svg>
  <svg className=\\"BlackFill\\">
    <path d=\\"M 0 250,\\\\n                         L 150 250,\\\\n                         A 100 100 0 0 1 350 250\\\\n                         L 500 250,\\\\n                         L 500 0\\\\n                         L 0 0\\" />
    <path d=\\"M 0 250,\\\\n                         L 150 250,\\\\n                         A 100 100 0 0 0 350 250\\\\n                         L 500 250,\\\\n                         L 500 500\\\\n                         L 0 500\\" />
  </svg>
  <Shadows openCoefficient={0.5} />
</svg>"
`;
